src/core/protocol-handlers/utils.ts

```diff
@@ -0,0 +1,196 @@
+/**
+ * Utility functions for protocol handlers
+ */
+
+import type { CalendarCollectionIndex } from '../../types/external-calendar';
+
+/**
+ * Normalize a location by adding .json extension if no extension is present
+ * Special handling for directory-like paths that should load index.json
+ */
+export function normalizeCalendarLocation(location: string): string {
+  // Remove fragment identifier for processing
+  const [basePath, fragment] = location.split('#');
+
+  // If location already has an extension, return as-is
+  if (hasFileExtension(basePath)) {
+    return location;
+  }
+
+  // Check if this looks like a directory path
+  if (isDirectoryPath(basePath)) {
+    // For directory paths, append index.json (preserve fragment)
+    return fragment ? `${basePath}/index.json#${fragment}` : `${basePath}/index.json`;
```

Comment:
I wonder if we could support variants within a calendar or external variants file as well as the index.json file.